spark/src/test/java/org/apache/iceberg/TestKryoSerialization.java

```diff
@@ -0,0 +1,130 @@
+      .build();
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Test
+  public void testDataFileKryoSerialization() throws Exception {
+    File data = temp.newFile();
+    Assert.assertTrue(data.delete());
+    Kryo kryo = new KryoSerializer(new SparkConf()).newKryo();
+
+    Map<Integer, Long> valueCounts = Maps.newHashMap();
+    valueCounts.put(1, 5L);
+    valueCounts.put(2, 3L);
+
+    Map<Integer, Long> nullValueCounts = Maps.newHashMap();
+    nullValueCounts.put(1, 0L);
+    nullValueCounts.put(2, 2L);
+
+    Map<Integer, ByteBuffer> lowerBounds = Maps.newHashMap();
+    lowerBounds.put(1, longToBuffer(0L));
+
+    Map<Integer, ByteBuffer> upperBounds = Maps.newHashMap();
+    upperBounds.put(1, longToBuffer(4L));
+
+    DataFile dataFile = DataFiles
+        .builder(PARTITION_SPEC)
+        .withPath("/path/to/data-1.parquet")
+        .withFileSizeInBytes(1234)
+        .withPartitionPath("date=2018-06-08")
+        .withMetrics(new Metrics(5L, null, valueCounts, nullValueCounts, lowerBounds, upperBounds))
+        .withSplitOffsets(ImmutableList.of(4L))
+        .build();
+
+    try (Output out = new Output(new FileOutputStream(data))) {
+      kryo.writeClassAndObject(out, dataFile);
+      kryo.writeClassAndObject(out, dataFile.copy());
+    }
+
+    try (Input in = new Input(new FileInputStream(data))) {
+      for (int i = 0; i < 2; i += 1) {
+        Object obj = kryo.readClassAndObject(in);
+        Assert.assertTrue("Should be a DataFile", obj instanceof DataFile);
+        DataFile result = (DataFile) obj;
+        Assert.assertEquals("Should match the serialized record path",
+            dataFile.path(), result.path());
+        Assert.assertEquals("Should match the serialized record format",
+            dataFile.format(), result.format());
+        Assert.assertEquals("Should match the serialized record partition",
+            dataFile.partition().get(0, Object.class), result.partition().get(0, Object.class));
+        Assert.assertEquals("Should match the serialized record count",
+            dataFile.recordCount(), result.recordCount());
+        Assert.assertEquals("Should match the serialized record size",
+            dataFile.fileSizeInBytes(), result.fileSizeInBytes());
+        Assert.assertEquals("Should match the serialized record value counts",
+            dataFile.valueCounts(), result.valueCounts());
+        Assert.assertEquals("Should match the serialized record null value counts",
+            dataFile.nullValueCounts(), result.nullValueCounts());
+        Assert.assertEquals("Should match the serialized record lower bounds",
+            dataFile.lowerBounds(), result.lowerBounds());
+        Assert.assertEquals("Should match the serialized record upper bounds",
+            dataFile.upperBounds(), result.upperBounds());
+        Assert.assertEquals("Should match the serialized record key metadata",
+            dataFile.keyMetadata(), result.keyMetadata());
```

Comment:
Did we set something in `keyMetadata`? There is some custom serialization logic to handle this field.

Reply:
No, this is empty.

Kryo can't handle ByteBuffers, but Spark's Kryo can. That's why this test uses Kryo provided by Spark. That's how this field is handled.